src/ItemsDialog.js

```diff
@@ -58,16 +65,102 @@ export const formatter = {
 export const MAX_HEIGHT = 500;
 
 const ItemsDialog = ({
-  title,
   onClose,
   open,
-  items,
   isLoading,
   onRowClick = noop,
+  mutator,
+  skippedItemId,
+  title,
+  instanceId,
 }) => {
+  const [areItemsBeingLoaded, setAreItemsBeingLoaded] = useState(false);
+  const [items, setItems] = useState([]);
   const { formatMessage } = useIntl();
-  const contentData = orderBy(items, 'requestQueue');
-  const count = items.length;
+
+  const fetchHoldings = () => {
+    const query = `instanceId==${instanceId}`;
+    mutator.holdings.reset();
+
+    return mutator.holdings.GET({ params: { query } });
+  };
+
+  const fetchItems = (holdings) => {
+    const query = holdings.map(h => `holdingsRecordId==${h.id}`).join(' or ');
+    mutator.items.reset();
+
+    return mutator.items.GET({ params: { query, limit: 1000 } });
+  };
+
+  const fetchRequests = async (itemsList) => {
+    // Split the list of items into small chunks to create a short enough query string
+    // that we can avoid a "414 Request URI Too Long" response from Okapi.
+    const CHUNK_SIZE = 40;
+
+    const chunkedItems = chunk(itemsList, CHUNK_SIZE);
+
+    const data = [];
+
+    for (const itemChunk of chunkedItems) {
+      let query = itemChunk.map(i => `itemId==${i.id}`).join(' or ');
+      query = `(${query}) and (status="Open")`;
+
+      mutator.requests.reset();
+      // eslint-disable-next-line no-await-in-loop
+      const result = await mutator.requests.GET({ params: { query, limit: 1000 } });
+
+      data.push(...result);
+    }
+
+    return data;
+  };
+
+  useLayoutEffect(() => {
+    const getItems = async () => {
+      setAreItemsBeingLoaded(true);
+
+      const holdings = await fetchHoldings();
+      let itemsList = await fetchItems(holdings);
+      const requests = await fetchRequests(itemsList);
+      const requestMap = countBy(requests, 'itemId');
+
+      itemsList = itemsList.map(item => ({ ...item, requestQueue: requestMap[item.id] || 0 }));
+
+      setAreItemsBeingLoaded(false);
+      setItems(itemsList);
+    };
+
+    if (open && instanceId) {
+      getItems();
+    }
+
+    return () => setItems([]);
+  }, [
+    open,
+    instanceId,
+  ]);
+
+  const contentData = useMemo(() => {
+    let resultItems = items;
+
+    if (skippedItemId) {
+      resultItems = resultItems
+        .filter(item => skippedItemId !== item.id);
+    }
+
+    // items with status available must go first
+    return resultItems.sort((a, b) => {
```

Comment:
@pavelspichonak I think you can shorter this with:

````js
return resultItems.sort(a => a?.status?.name === itemStatuses.AVAILABLE ? -1 : 1);
````

Reply:
Done. Thanks